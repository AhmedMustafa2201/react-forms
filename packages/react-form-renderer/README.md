# React Form Renderer

Rendering forms from data structures

## Documentation: http://data-driven-forms.org/
