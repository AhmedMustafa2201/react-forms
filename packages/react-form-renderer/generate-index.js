const glob = require('glob');
const path = require('path');
const fse = require('fs-extra');

const src = path.resolve(__dirname, './src');

function camelize(str) {
  const arr = str.split('-');
  const capital = arr.map((item, index) => (index ? item.charAt(0).toUpperCase() + item.slice(1).toLowerCase() : item.toLowerCase()));
  return capital.join('');
}

function generateIndex() {
  const directories = glob
    .sync(`${src}/*/`)
    .filter((name) => !name.includes('/tests/') && !name.includes('/common/'))
    .map((path) =>
      path
        .replace(/\/$/, '')
        .split('/')
        .pop()
    );
  const content = `${directories.map(
    (dir) => `export { default as ${camelize(dir)} } from './${dir}';
`
  )}`.replace(/,/g, '');

  return fse.writeFileSync(`${src}/index.js`, content);
}

function run() {
  try {
    generateIndex();
  } catch (err) {
    console.error(err);
    process.exit(1);
  }
}

run();
