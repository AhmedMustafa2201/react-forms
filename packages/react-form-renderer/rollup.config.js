import nodeResolve from 'rollup-plugin-node-resolve';
import commonjs from 'rollup-plugin-commonjs';
import babel from 'rollup-plugin-babel';
import replace from 'rollup-plugin-replace';
import nodeGlobals from 'rollup-plugin-node-globals';
import { terser } from 'rollup-plugin-terser';
import { sizeSnapshot } from 'rollup-plugin-size-snapshot';

const globals = {
  react: 'React',
  'react-dom': 'ReactDOM',
};

const babelOptions = {
  exclude: /node_modules/,
  runtimeHelpers: true,
  configFile: '../../babel.config.js',
};

const commonjsOptions = {
  ignoreGlobal: true,
  include: /node_modules/,
  namedExports: {
    './src/form-renderer/helpers': [ 'composeValidators' ],
  },
};

function onwarn(warning) {
  throw Error(warning.message);
}

export default [{
  onwarn,
  input: './src/index.js',
  output: {
    file: './dist/index.js',
    format: 'umd',
    name: '@data-driven-forms/react-form-renderer',
    exports: 'named',
    globals,
  },
  external: Object.keys(globals),
  plugins: [
    nodeResolve(),
    babel(babelOptions),
    commonjs(commonjsOptions),
    nodeGlobals(), // Wait for https://github.com/cssinjs/jss/pull/893
    replace({ 'process.env.NODE_ENV': JSON.stringify('production') }),
    sizeSnapshot({ snapshotPath: 'size-snapshot.json' }),
    terser({
      keep_classnames: true,
      keep_fnames: true,
    }),
  ],
}];
