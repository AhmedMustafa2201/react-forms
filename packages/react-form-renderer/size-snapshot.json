{
  "dist/index.js": {
    "bundled": 246500,
    "minified": 87931,
    "gzipped": 26546
  }
}
