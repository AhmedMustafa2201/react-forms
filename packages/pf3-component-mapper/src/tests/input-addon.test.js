import React from 'react';
import toJson from 'enzyme-to-json';
import { mount } from 'enzyme';
import FormRenderer from '@data-driven-forms/react-form-renderer';
import { componentTypes as components } from '@data-driven-forms/react-form-renderer';
import { formFieldsMapper, layoutMapper } from '../';

const schema = {
  fields: [
    {
      name: 'text_box_21',
      label: 'Text Box With Input Addon',
      title: 'Text Box With Input Addon',
      component: components.TEXT_FIELD
    }
  ]
};

describe('<Input Addon>', () => {
  it('should render single before input addon correctly', () => {
    const inputAddon = {
      before: {
        fields: [
          {
            component: components.PLAIN_TEXT,
            label: '-',
            name: 'name1',
            variant: 'span'
          }
        ]
      }
    };
    const wrapper = mount(
      <FormRenderer
        schema={{ ...schema, fields: [{ ...schema.fields[0], inputAddon }] }}
        formFieldsMapper={formFieldsMapper}
        layoutMapper={layoutMapper}
        onCancel={() => {}}
        showFormControls={false}
        onSubmit={jest.fn()}
      />
    );
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it('should render single button before input addon correctly', () => {
    const inputAddon = {
      before: {
        fields: [
          {
            component: components.INPUT_ADDON_BUTTON_GROUP,
            name: 'i-a-g-2',
            fields: [
              {
                component: components.BUTTON,
                label: 'Set 2',
                name: 'set2'
              }
            ]
          }
        ]
      }
    };
    const wrapper = mount(
      <FormRenderer
        schema={{ ...schema, fields: [{ ...schema.fields[0], inputAddon }] }}
        formFieldsMapper={formFieldsMapper}
        layoutMapper={layoutMapper}
        onCancel={() => {}}
        showFormControls={false}
        onSubmit={jest.fn()}
      />
    );
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it('should render single after input addon correctly', () => {
    const inputAddon = {
      after: {
        fields: [
          {
            component: components.PLAIN_TEXT,
            label: '-',
            name: 'name1',
            variant: 'span'
          }
        ]
      }
    };
    const wrapper = mount(
      <FormRenderer
        schema={{ ...schema, fields: [{ ...schema.fields[0], inputAddon }] }}
        formFieldsMapper={formFieldsMapper}
        layoutMapper={layoutMapper}
        onCancel={() => {}}
        showFormControls={false}
        onSubmit={jest.fn()}
      />
    );
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it('should render single button after input addon correctly', () => {
    const inputAddon = {
      after: {
        fields: [
          {
            component: components.BUTTON,
            label: 'Set 3',
            name: 'set3'
          }
        ]
      }
    };
    const wrapper = mount(
      <FormRenderer
        schema={{ ...schema, fields: [{ ...schema.fields[0], inputAddon }] }}
        formFieldsMapper={formFieldsMapper}
        layoutMapper={layoutMapper}
        onCancel={() => {}}
        showFormControls={false}
        onSubmit={jest.fn()}
      />
    );
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it('should render group button before input addon correctly', () => {
    const inputAddon = {
      before: {
        fields: [
          {
            component: components.INPUT_ADDON_BUTTON_GROUP,
            name: 'i-a-g-2',
            fields: [
              {
                component: components.BUTTON,
                label: 'Set 1',
                name: 'set1',
                onClick: () => {}
              },
              {
                component: components.BUTTON,
                label: 'Set 2',
                name: 'set2'
              }
            ]
          }
        ]
      }
    };
    const wrapper = mount(
      <FormRenderer
        schema={{ ...schema, fields: [{ ...schema.fields[0], inputAddon }] }}
        formFieldsMapper={formFieldsMapper}
        layoutMapper={layoutMapper}
        onCancel={() => {}}
        showFormControls={false}
        onSubmit={jest.fn()}
      />
    );
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it('should render group mixed before input addon correctly', () => {
    const inputAddon = {
      before: {
        fields: [
          {
            component: components.INPUT_ADDON_GROUP,
            name: 'i-a-g-2',
            fields: [
              {
                component: components.PLAIN_TEXT,
                label: '-',
                name: 'name1',
                variant: 'span'
              }
            ]
          },
          {
            component: components.INPUT_ADDON_BUTTON_GROUP,
            name: 'i-a-g-3',
            fields: [
              {
                component: components.BUTTON,
                label: 'Set 3',
                name: 'set3'
              }
            ]
          }
        ]
      }
    };
    const wrapper = mount(
      <FormRenderer
        schema={{ ...schema, fields: [{ ...schema.fields[0], inputAddon }] }}
        formFieldsMapper={formFieldsMapper}
        layoutMapper={layoutMapper}
        onCancel={() => {}}
        showFormControls={false}
        onSubmit={jest.fn()}
      />
    );
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it('should render the ultimate input addon correctly', () => {
    const inputAddon = {
      before: {
        fields: [
          {
            component: components.INPUT_ADDON_BUTTON_GROUP,
            name: 'i-a-g-2',
            fields: [
              {
                component: components.BUTTON,
                label: 'Set 1',
                name: 'set1',
                onClick: () => {}
              },
              {
                component: components.BUTTON,
                label: 'Set 2',
                name: 'set2'
              }
            ]
          }
        ]
      },
      after: {
        fields: [
          {
            component: components.INPUT_ADDON_GROUP,
            name: 'i-a-g-2',
            fields: [
              {
                component: components.PLAIN_TEXT,
                label: '-',
                name: 'name1',
                variant: 'span'
              }
            ]
          },
          {
            component: components.INPUT_ADDON_BUTTON_GROUP,
            name: 'i-a-g-3',
            fields: [
              {
                component: components.BUTTON,
                label: 'Set 3',
                name: 'set3'
              }
            ]
          }
        ]
      }
    };
    const wrapper = mount(
      <FormRenderer
        schema={{ ...schema, fields: [{ ...schema.fields[0], inputAddon }] }}
        formFieldsMapper={formFieldsMapper}
        layoutMapper={layoutMapper}
        onCancel={() => {}}
        showFormControls={false}
        onSubmit={jest.fn()}
      />
    );
    expect(toJson(wrapper)).toMatchSnapshot();
  });
});
