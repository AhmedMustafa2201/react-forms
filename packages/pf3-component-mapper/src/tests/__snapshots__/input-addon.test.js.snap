// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Input Addon> should render group button before input addon correctly 1`] = `
<FormRenderer
  clearOnUnmount={false}
  componentMapper={
    Object {
      "button": [Function],
      "checkbox": [Function],
      "date-picker": [Function],
      "input-addon-button-group": [Function],
      "input-addon-group": [Function],
      "plain-text": [Function],
      "radio": [Function],
      "select-field": [Function],
      "sub-form": [Function],
      "switch-field": [Function],
      "tabs": [Function],
      "text-field": [Function],
      "textarea-field": [Function],
      "time-picker": [Function],
      "wizard": [Function],
    }
  }
  initialValues={Object {}}
  onCancel={[Function]}
  onSubmit={[MockFunction]}
  schema={
    Object {
      "fields": Array [
        Object {
          "component": "text-field",
          "inputAddon": Object {
            "before": Object {
              "fields": Array [
                Object {
                  "component": "input-addon-button-group",
                  "fields": Array [
                    Object {
                      "component": "button",
                      "label": "Set 1",
                      "name": "set1",
                      "onClick": [Function],
                    },
                    Object {
                      "component": "button",
                      "label": "Set 2",
                      "name": "set2",
                    },
                  ],
                  "name": "i-a-g-2",
                },
              ],
            },
          },
          "label": "Text Box With Input Addon",
          "name": "text_box_21",
          "title": "Text Box With Input Addon",
        },
      ],
    }
  }
  showFormControls={false}
>
  <ReactFinalForm
    decorators={
      Array [
        [Function],
      ]
    }
    initialValues={Object {}}
    mutators={
      Object {
        "concat": [Function],
        "insert": [Function],
        "move": [Function],
        "pop": [Function],
        "push": [Function],
        "remove": [Function],
        "removeBatch": [Function],
        "shift": [Function],
        "swap": [Function],
        "unshift": [Function],
        "update": [Function],
      }
    }
    onSubmit={[MockFunction]}
    render={[Function]}
    subscription={
      Object {
        "pristine": true,
        "submitting": true,
        "valid": true,
      }
    }
  >
    <Component
      formFields={
        Array [
          <SingleField
            component="text-field"
            inputAddon={
              Object {
                "before": Object {
                  "fields": Array [
                    Object {
                      "component": "input-addon-button-group",
                      "fields": Array [
                        Object {
                          "component": "button",
                          "label": "Set 1",
                          "name": "set1",
                          "onClick": [Function],
                        },
                        Object {
                          "component": "button",
                          "label": "Set 2",
                          "name": "set2",
                        },
                      ],
                      "name": "i-a-g-2",
                    },
                  ],
                },
              }
            }
            label="Text Box With Input Addon"
            name="text_box_21"
            title="Text Box With Input Addon"
          />,
        ]
      }
      schema={
        Object {
          "fields": Array [
            Object {
              "component": "text-field",
              "inputAddon": Object {
                "before": Object {
                  "fields": Array [
                    Object {
                      "component": "input-addon-button-group",
                      "fields": Array [
                        Object {
                          "component": "button",
                          "label": "Set 1",
                          "name": "set1",
                          "onClick": [Function],
                        },
                        Object {
                          "component": "button",
                          "label": "Set 2",
                          "name": "set2",
                        },
                      ],
                      "name": "i-a-g-2",
                    },
                  ],
                },
              },
              "label": "Text Box With Input Addon",
              "name": "text_box_21",
              "title": "Text Box With Input Addon",
            },
          ],
        }
      }
    >
      <div>
        FormRenderer is missing 'formTemplate' prop: ({formFields, schema}) =&gt; &lt;FormTemplate {...} /&gt;
      </div>
    </Component>
  </ReactFinalForm>
</FormRenderer>
`;

exports[`<Input Addon> should render group mixed before input addon correctly 1`] = `
<FormRenderer
  clearOnUnmount={false}
  componentMapper={
    Object {
      "button": [Function],
      "checkbox": [Function],
      "date-picker": [Function],
      "input-addon-button-group": [Function],
      "input-addon-group": [Function],
      "plain-text": [Function],
      "radio": [Function],
      "select-field": [Function],
      "sub-form": [Function],
      "switch-field": [Function],
      "tabs": [Function],
      "text-field": [Function],
      "textarea-field": [Function],
      "time-picker": [Function],
      "wizard": [Function],
    }
  }
  initialValues={Object {}}
  onCancel={[Function]}
  onSubmit={[MockFunction]}
  schema={
    Object {
      "fields": Array [
        Object {
          "component": "text-field",
          "inputAddon": Object {
            "before": Object {
              "fields": Array [
                Object {
                  "component": "input-addon-group",
                  "fields": Array [
                    Object {
                      "component": "plain-text",
                      "label": "-",
                      "name": "name1",
                      "variant": "span",
                    },
                  ],
                  "name": "i-a-g-2",
                },
                Object {
                  "component": "input-addon-button-group",
                  "fields": Array [
                    Object {
                      "component": "button",
                      "label": "Set 3",
                      "name": "set3",
                    },
                  ],
                  "name": "i-a-g-3",
                },
              ],
            },
          },
          "label": "Text Box With Input Addon",
          "name": "text_box_21",
          "title": "Text Box With Input Addon",
        },
      ],
    }
  }
  showFormControls={false}
>
  <ReactFinalForm
    decorators={
      Array [
        [Function],
      ]
    }
    initialValues={Object {}}
    mutators={
      Object {
        "concat": [Function],
        "insert": [Function],
        "move": [Function],
        "pop": [Function],
        "push": [Function],
        "remove": [Function],
        "removeBatch": [Function],
        "shift": [Function],
        "swap": [Function],
        "unshift": [Function],
        "update": [Function],
      }
    }
    onSubmit={[MockFunction]}
    render={[Function]}
    subscription={
      Object {
        "pristine": true,
        "submitting": true,
        "valid": true,
      }
    }
  >
    <Component
      formFields={
        Array [
          <SingleField
            component="text-field"
            inputAddon={
              Object {
                "before": Object {
                  "fields": Array [
                    Object {
                      "component": "input-addon-group",
                      "fields": Array [
                        Object {
                          "component": "plain-text",
                          "label": "-",
                          "name": "name1",
                          "variant": "span",
                        },
                      ],
                      "name": "i-a-g-2",
                    },
                    Object {
                      "component": "input-addon-button-group",
                      "fields": Array [
                        Object {
                          "component": "button",
                          "label": "Set 3",
                          "name": "set3",
                        },
                      ],
                      "name": "i-a-g-3",
                    },
                  ],
                },
              }
            }
            label="Text Box With Input Addon"
            name="text_box_21"
            title="Text Box With Input Addon"
          />,
        ]
      }
      schema={
        Object {
          "fields": Array [
            Object {
              "component": "text-field",
              "inputAddon": Object {
                "before": Object {
                  "fields": Array [
                    Object {
                      "component": "input-addon-group",
                      "fields": Array [
                        Object {
                          "component": "plain-text",
                          "label": "-",
                          "name": "name1",
                          "variant": "span",
                        },
                      ],
                      "name": "i-a-g-2",
                    },
                    Object {
                      "component": "input-addon-button-group",
                      "fields": Array [
                        Object {
                          "component": "button",
                          "label": "Set 3",
                          "name": "set3",
                        },
                      ],
                      "name": "i-a-g-3",
                    },
                  ],
                },
              },
              "label": "Text Box With Input Addon",
              "name": "text_box_21",
              "title": "Text Box With Input Addon",
            },
          ],
        }
      }
    >
      <div>
        FormRenderer is missing 'formTemplate' prop: ({formFields, schema}) =&gt; &lt;FormTemplate {...} /&gt;
      </div>
    </Component>
  </ReactFinalForm>
</FormRenderer>
`;

exports[`<Input Addon> should render single after input addon correctly 1`] = `
<FormRenderer
  clearOnUnmount={false}
  componentMapper={
    Object {
      "button": [Function],
      "checkbox": [Function],
      "date-picker": [Function],
      "input-addon-button-group": [Function],
      "input-addon-group": [Function],
      "plain-text": [Function],
      "radio": [Function],
      "select-field": [Function],
      "sub-form": [Function],
      "switch-field": [Function],
      "tabs": [Function],
      "text-field": [Function],
      "textarea-field": [Function],
      "time-picker": [Function],
      "wizard": [Function],
    }
  }
  initialValues={Object {}}
  onCancel={[Function]}
  onSubmit={[MockFunction]}
  schema={
    Object {
      "fields": Array [
        Object {
          "component": "text-field",
          "inputAddon": Object {
            "after": Object {
              "fields": Array [
                Object {
                  "component": "plain-text",
                  "label": "-",
                  "name": "name1",
                  "variant": "span",
                },
              ],
            },
          },
          "label": "Text Box With Input Addon",
          "name": "text_box_21",
          "title": "Text Box With Input Addon",
        },
      ],
    }
  }
  showFormControls={false}
>
  <ReactFinalForm
    decorators={
      Array [
        [Function],
      ]
    }
    initialValues={Object {}}
    mutators={
      Object {
        "concat": [Function],
        "insert": [Function],
        "move": [Function],
        "pop": [Function],
        "push": [Function],
        "remove": [Function],
        "removeBatch": [Function],
        "shift": [Function],
        "swap": [Function],
        "unshift": [Function],
        "update": [Function],
      }
    }
    onSubmit={[MockFunction]}
    render={[Function]}
    subscription={
      Object {
        "pristine": true,
        "submitting": true,
        "valid": true,
      }
    }
  >
    <Component
      formFields={
        Array [
          <SingleField
            component="text-field"
            inputAddon={
              Object {
                "after": Object {
                  "fields": Array [
                    Object {
                      "component": "plain-text",
                      "label": "-",
                      "name": "name1",
                      "variant": "span",
                    },
                  ],
                },
              }
            }
            label="Text Box With Input Addon"
            name="text_box_21"
            title="Text Box With Input Addon"
          />,
        ]
      }
      schema={
        Object {
          "fields": Array [
            Object {
              "component": "text-field",
              "inputAddon": Object {
                "after": Object {
                  "fields": Array [
                    Object {
                      "component": "plain-text",
                      "label": "-",
                      "name": "name1",
                      "variant": "span",
                    },
                  ],
                },
              },
              "label": "Text Box With Input Addon",
              "name": "text_box_21",
              "title": "Text Box With Input Addon",
            },
          ],
        }
      }
    >
      <div>
        FormRenderer is missing 'formTemplate' prop: ({formFields, schema}) =&gt; &lt;FormTemplate {...} /&gt;
      </div>
    </Component>
  </ReactFinalForm>
</FormRenderer>
`;

exports[`<Input Addon> should render single before input addon correctly 1`] = `
<FormRenderer
  clearOnUnmount={false}
  componentMapper={
    Object {
      "button": [Function],
      "checkbox": [Function],
      "date-picker": [Function],
      "input-addon-button-group": [Function],
      "input-addon-group": [Function],
      "plain-text": [Function],
      "radio": [Function],
      "select-field": [Function],
      "sub-form": [Function],
      "switch-field": [Function],
      "tabs": [Function],
      "text-field": [Function],
      "textarea-field": [Function],
      "time-picker": [Function],
      "wizard": [Function],
    }
  }
  initialValues={Object {}}
  onCancel={[Function]}
  onSubmit={[MockFunction]}
  schema={
    Object {
      "fields": Array [
        Object {
          "component": "text-field",
          "inputAddon": Object {
            "before": Object {
              "fields": Array [
                Object {
                  "component": "plain-text",
                  "label": "-",
                  "name": "name1",
                  "variant": "span",
                },
              ],
            },
          },
          "label": "Text Box With Input Addon",
          "name": "text_box_21",
          "title": "Text Box With Input Addon",
        },
      ],
    }
  }
  showFormControls={false}
>
  <ReactFinalForm
    decorators={
      Array [
        [Function],
      ]
    }
    initialValues={Object {}}
    mutators={
      Object {
        "concat": [Function],
        "insert": [Function],
        "move": [Function],
        "pop": [Function],
        "push": [Function],
        "remove": [Function],
        "removeBatch": [Function],
        "shift": [Function],
        "swap": [Function],
        "unshift": [Function],
        "update": [Function],
      }
    }
    onSubmit={[MockFunction]}
    render={[Function]}
    subscription={
      Object {
        "pristine": true,
        "submitting": true,
        "valid": true,
      }
    }
  >
    <Component
      formFields={
        Array [
          <SingleField
            component="text-field"
            inputAddon={
              Object {
                "before": Object {
                  "fields": Array [
                    Object {
                      "component": "plain-text",
                      "label": "-",
                      "name": "name1",
                      "variant": "span",
                    },
                  ],
                },
              }
            }
            label="Text Box With Input Addon"
            name="text_box_21"
            title="Text Box With Input Addon"
          />,
        ]
      }
      schema={
        Object {
          "fields": Array [
            Object {
              "component": "text-field",
              "inputAddon": Object {
                "before": Object {
                  "fields": Array [
                    Object {
                      "component": "plain-text",
                      "label": "-",
                      "name": "name1",
                      "variant": "span",
                    },
                  ],
                },
              },
              "label": "Text Box With Input Addon",
              "name": "text_box_21",
              "title": "Text Box With Input Addon",
            },
          ],
        }
      }
    >
      <div>
        FormRenderer is missing 'formTemplate' prop: ({formFields, schema}) =&gt; &lt;FormTemplate {...} /&gt;
      </div>
    </Component>
  </ReactFinalForm>
</FormRenderer>
`;

exports[`<Input Addon> should render single button after input addon correctly 1`] = `
<FormRenderer
  clearOnUnmount={false}
  componentMapper={
    Object {
      "button": [Function],
      "checkbox": [Function],
      "date-picker": [Function],
      "input-addon-button-group": [Function],
      "input-addon-group": [Function],
      "plain-text": [Function],
      "radio": [Function],
      "select-field": [Function],
      "sub-form": [Function],
      "switch-field": [Function],
      "tabs": [Function],
      "text-field": [Function],
      "textarea-field": [Function],
      "time-picker": [Function],
      "wizard": [Function],
    }
  }
  initialValues={Object {}}
  onCancel={[Function]}
  onSubmit={[MockFunction]}
  schema={
    Object {
      "fields": Array [
        Object {
          "component": "text-field",
          "inputAddon": Object {
            "after": Object {
              "fields": Array [
                Object {
                  "component": "button",
                  "label": "Set 3",
                  "name": "set3",
                },
              ],
            },
          },
          "label": "Text Box With Input Addon",
          "name": "text_box_21",
          "title": "Text Box With Input Addon",
        },
      ],
    }
  }
  showFormControls={false}
>
  <ReactFinalForm
    decorators={
      Array [
        [Function],
      ]
    }
    initialValues={Object {}}
    mutators={
      Object {
        "concat": [Function],
        "insert": [Function],
        "move": [Function],
        "pop": [Function],
        "push": [Function],
        "remove": [Function],
        "removeBatch": [Function],
        "shift": [Function],
        "swap": [Function],
        "unshift": [Function],
        "update": [Function],
      }
    }
    onSubmit={[MockFunction]}
    render={[Function]}
    subscription={
      Object {
        "pristine": true,
        "submitting": true,
        "valid": true,
      }
    }
  >
    <Component
      formFields={
        Array [
          <SingleField
            component="text-field"
            inputAddon={
              Object {
                "after": Object {
                  "fields": Array [
                    Object {
                      "component": "button",
                      "label": "Set 3",
                      "name": "set3",
                    },
                  ],
                },
              }
            }
            label="Text Box With Input Addon"
            name="text_box_21"
            title="Text Box With Input Addon"
          />,
        ]
      }
      schema={
        Object {
          "fields": Array [
            Object {
              "component": "text-field",
              "inputAddon": Object {
                "after": Object {
                  "fields": Array [
                    Object {
                      "component": "button",
                      "label": "Set 3",
                      "name": "set3",
                    },
                  ],
                },
              },
              "label": "Text Box With Input Addon",
              "name": "text_box_21",
              "title": "Text Box With Input Addon",
            },
          ],
        }
      }
    >
      <div>
        FormRenderer is missing 'formTemplate' prop: ({formFields, schema}) =&gt; &lt;FormTemplate {...} /&gt;
      </div>
    </Component>
  </ReactFinalForm>
</FormRenderer>
`;

exports[`<Input Addon> should render single button before input addon correctly 1`] = `
<FormRenderer
  clearOnUnmount={false}
  componentMapper={
    Object {
      "button": [Function],
      "checkbox": [Function],
      "date-picker": [Function],
      "input-addon-button-group": [Function],
      "input-addon-group": [Function],
      "plain-text": [Function],
      "radio": [Function],
      "select-field": [Function],
      "sub-form": [Function],
      "switch-field": [Function],
      "tabs": [Function],
      "text-field": [Function],
      "textarea-field": [Function],
      "time-picker": [Function],
      "wizard": [Function],
    }
  }
  initialValues={Object {}}
  onCancel={[Function]}
  onSubmit={[MockFunction]}
  schema={
    Object {
      "fields": Array [
        Object {
          "component": "text-field",
          "inputAddon": Object {
            "before": Object {
              "fields": Array [
                Object {
                  "component": "input-addon-button-group",
                  "fields": Array [
                    Object {
                      "component": "button",
                      "label": "Set 2",
                      "name": "set2",
                    },
                  ],
                  "name": "i-a-g-2",
                },
              ],
            },
          },
          "label": "Text Box With Input Addon",
          "name": "text_box_21",
          "title": "Text Box With Input Addon",
        },
      ],
    }
  }
  showFormControls={false}
>
  <ReactFinalForm
    decorators={
      Array [
        [Function],
      ]
    }
    initialValues={Object {}}
    mutators={
      Object {
        "concat": [Function],
        "insert": [Function],
        "move": [Function],
        "pop": [Function],
        "push": [Function],
        "remove": [Function],
        "removeBatch": [Function],
        "shift": [Function],
        "swap": [Function],
        "unshift": [Function],
        "update": [Function],
      }
    }
    onSubmit={[MockFunction]}
    render={[Function]}
    subscription={
      Object {
        "pristine": true,
        "submitting": true,
        "valid": true,
      }
    }
  >
    <Component
      formFields={
        Array [
          <SingleField
            component="text-field"
            inputAddon={
              Object {
                "before": Object {
                  "fields": Array [
                    Object {
                      "component": "input-addon-button-group",
                      "fields": Array [
                        Object {
                          "component": "button",
                          "label": "Set 2",
                          "name": "set2",
                        },
                      ],
                      "name": "i-a-g-2",
                    },
                  ],
                },
              }
            }
            label="Text Box With Input Addon"
            name="text_box_21"
            title="Text Box With Input Addon"
          />,
        ]
      }
      schema={
        Object {
          "fields": Array [
            Object {
              "component": "text-field",
              "inputAddon": Object {
                "before": Object {
                  "fields": Array [
                    Object {
                      "component": "input-addon-button-group",
                      "fields": Array [
                        Object {
                          "component": "button",
                          "label": "Set 2",
                          "name": "set2",
                        },
                      ],
                      "name": "i-a-g-2",
                    },
                  ],
                },
              },
              "label": "Text Box With Input Addon",
              "name": "text_box_21",
              "title": "Text Box With Input Addon",
            },
          ],
        }
      }
    >
      <div>
        FormRenderer is missing 'formTemplate' prop: ({formFields, schema}) =&gt; &lt;FormTemplate {...} /&gt;
      </div>
    </Component>
  </ReactFinalForm>
</FormRenderer>
`;

exports[`<Input Addon> should render the ultimate input addon correctly 1`] = `
<FormRenderer
  clearOnUnmount={false}
  componentMapper={
    Object {
      "button": [Function],
      "checkbox": [Function],
      "date-picker": [Function],
      "input-addon-button-group": [Function],
      "input-addon-group": [Function],
      "plain-text": [Function],
      "radio": [Function],
      "select-field": [Function],
      "sub-form": [Function],
      "switch-field": [Function],
      "tabs": [Function],
      "text-field": [Function],
      "textarea-field": [Function],
      "time-picker": [Function],
      "wizard": [Function],
    }
  }
  initialValues={Object {}}
  onCancel={[Function]}
  onSubmit={[MockFunction]}
  schema={
    Object {
      "fields": Array [
        Object {
          "component": "text-field",
          "inputAddon": Object {
            "after": Object {
              "fields": Array [
                Object {
                  "component": "input-addon-group",
                  "fields": Array [
                    Object {
                      "component": "plain-text",
                      "label": "-",
                      "name": "name1",
                      "variant": "span",
                    },
                  ],
                  "name": "i-a-g-2",
                },
                Object {
                  "component": "input-addon-button-group",
                  "fields": Array [
                    Object {
                      "component": "button",
                      "label": "Set 3",
                      "name": "set3",
                    },
                  ],
                  "name": "i-a-g-3",
                },
              ],
            },
            "before": Object {
              "fields": Array [
                Object {
                  "component": "input-addon-button-group",
                  "fields": Array [
                    Object {
                      "component": "button",
                      "label": "Set 1",
                      "name": "set1",
                      "onClick": [Function],
                    },
                    Object {
                      "component": "button",
                      "label": "Set 2",
                      "name": "set2",
                    },
                  ],
                  "name": "i-a-g-2",
                },
              ],
            },
          },
          "label": "Text Box With Input Addon",
          "name": "text_box_21",
          "title": "Text Box With Input Addon",
        },
      ],
    }
  }
  showFormControls={false}
>
  <ReactFinalForm
    decorators={
      Array [
        [Function],
      ]
    }
    initialValues={Object {}}
    mutators={
      Object {
        "concat": [Function],
        "insert": [Function],
        "move": [Function],
        "pop": [Function],
        "push": [Function],
        "remove": [Function],
        "removeBatch": [Function],
        "shift": [Function],
        "swap": [Function],
        "unshift": [Function],
        "update": [Function],
      }
    }
    onSubmit={[MockFunction]}
    render={[Function]}
    subscription={
      Object {
        "pristine": true,
        "submitting": true,
        "valid": true,
      }
    }
  >
    <Component
      formFields={
        Array [
          <SingleField
            component="text-field"
            inputAddon={
              Object {
                "after": Object {
                  "fields": Array [
                    Object {
                      "component": "input-addon-group",
                      "fields": Array [
                        Object {
                          "component": "plain-text",
                          "label": "-",
                          "name": "name1",
                          "variant": "span",
                        },
                      ],
                      "name": "i-a-g-2",
                    },
                    Object {
                      "component": "input-addon-button-group",
                      "fields": Array [
                        Object {
                          "component": "button",
                          "label": "Set 3",
                          "name": "set3",
                        },
                      ],
                      "name": "i-a-g-3",
                    },
                  ],
                },
                "before": Object {
                  "fields": Array [
                    Object {
                      "component": "input-addon-button-group",
                      "fields": Array [
                        Object {
                          "component": "button",
                          "label": "Set 1",
                          "name": "set1",
                          "onClick": [Function],
                        },
                        Object {
                          "component": "button",
                          "label": "Set 2",
                          "name": "set2",
                        },
                      ],
                      "name": "i-a-g-2",
                    },
                  ],
                },
              }
            }
            label="Text Box With Input Addon"
            name="text_box_21"
            title="Text Box With Input Addon"
          />,
        ]
      }
      schema={
        Object {
          "fields": Array [
            Object {
              "component": "text-field",
              "inputAddon": Object {
                "after": Object {
                  "fields": Array [
                    Object {
                      "component": "input-addon-group",
                      "fields": Array [
                        Object {
                          "component": "plain-text",
                          "label": "-",
                          "name": "name1",
                          "variant": "span",
                        },
                      ],
                      "name": "i-a-g-2",
                    },
                    Object {
                      "component": "input-addon-button-group",
                      "fields": Array [
                        Object {
                          "component": "button",
                          "label": "Set 3",
                          "name": "set3",
                        },
                      ],
                      "name": "i-a-g-3",
                    },
                  ],
                },
                "before": Object {
                  "fields": Array [
                    Object {
                      "component": "input-addon-button-group",
                      "fields": Array [
                        Object {
                          "component": "button",
                          "label": "Set 1",
                          "name": "set1",
                          "onClick": [Function],
                        },
                        Object {
                          "component": "button",
                          "label": "Set 2",
                          "name": "set2",
                        },
                      ],
                      "name": "i-a-g-2",
                    },
                  ],
                },
              },
              "label": "Text Box With Input Addon",
              "name": "text_box_21",
              "title": "Text Box With Input Addon",
            },
          ],
        }
      }
    >
      <div>
        FormRenderer is missing 'formTemplate' prop: ({formFields, schema}) =&gt; &lt;FormTemplate {...} /&gt;
      </div>
    </Component>
  </ReactFinalForm>
</FormRenderer>
`;
